jlm/rvsdg/traverser.hpp

```diff
@@ -24,203 +23,206 @@ class Region;
- * created nodes might also be traversed. The main usage of the topdown traverser is for replacing
- * subgraphs in the already visited part of a region.
- *
- * The topdown traverser associates three distinct states with any node in the region throughout
- * traversal:
- *
- * 1. <b>ahead</b>: Nodes that have not been visited yet and are not yet marked for visitation.
- * 2. <b>frontier</b>: Nodes that are marked for visitation.
- * 3. <b>behind</b>: Nodes that were already visited.
- *
- * All nodes are by default in state <em>ahead</em>. The topdown_traverser() constructor associates
- * the <em>frontier</em> state with the top-most nodes in the region, <em>i.e.</em>, all nodes that
- * have no inputs or only region arguments as origins. The next() method can then be used to
- * traverse these <em>frontier</em> nodes. Before a <em>frontier</em> node is returned by next(), it
- * is marked as <em>behind</em> and all nodes that depend on its outputs are transferred from the
- * <em>ahead</em> state to state <em>frontier</em>. The repeated invocation of next() traverses all
- * nodes in the region.
- *
- * A newly created node is marked as <em>behind</em> iff all the nodes' predecessors are marked as
- * behind. Otherwise, it is marked as <em>frontier</em>.
- *
- * An alternative to traversing all nodes using next() is the utilization of begin() and end().
- */
-class TopDownTraverser final
-{
-public:
-  ~TopDownTraverser() noexcept;
+  ~TopDownTraverserGeneric() noexcept;
 
-  explicit TopDownTraverser(Region * region);
+  explicit TopDownTraverserGeneric(RegionType * region);
 
-  Node *
+  NodeType *
   next();
 
-  typedef detail::TraverserIterator<TopDownTraverser> iterator;
-  typedef Node * value_type;
-
-  inline iterator
+  [[nodiscard]] iterator
   begin()
   {
     return iterator(this, next());
   }
 
-  inline iterator
+  [[nodiscard]] iterator
   end()
   {
     return iterator(this, nullptr);
   }
 
 private:
   bool
-  predecessors_visited(const Node * node) noexcept;
-
-  void
-  node_create(Node * node);
-
-  void
-  input_change(Input * in, Output * old_origin, Output * new_origin);
-
-  Region & region_;
-  TraversalTracker tracker_;
-  std::vector<jlm::util::Callback> callbacks_;
+  allPredecessorsVisited(const Node & node) const noexcept;
+
+  /**
+   * If the frontier is empty, while some nodes have yet to be visited,
+   * this method builds a new frontier.
+   * @return true if there is a frontier, false is all nodes have been visited.
+   */
+  [[nodiscard]] bool
+  prepareFrontier();
+
+  /**
+   * Extracts a node from the frontier.
+   * In non-const traversals, it double checks that the node is actually ready to be visited.
+   * @returns the popped node, or nullptr if all nodes have been visited.
+   */
+  [[nodiscard]] NodeType *
+  popFrontier();
+
+  // The set of nodes that are yet to be visited
+  util::HashSet<NodeType *> unvisited_;
```

Comment:
Keeping track of the nodes that have not been visited yet is expensive, as it effectively means you have to go through all the nodes in the region twice. Once, to fill this set, and then once the actual traversal. Would it not be possible to inverse it and keep track of the visited nodes, and all other nodes are by default in the unvisited state?